### PR TITLE
Fix local provider not found when bean initialize

### LINKED ISSUE #13672: Unable to find providers if switch to injvm protocol
### Environment

* Dubbo version: 3.0+
* Java version: 17

### 服务配置如下
```
dubbo:
  application:
    logger: slf4j
    qos-enable: false
  consumer:
    check: false
    scope: local
  protocol:
    accepts: 200
    threadpool: cached
    threads: 200
    name: injvm
  provider:
    scope: local
  registry:
    address: N/A
  scan:
    base-packages: com.freedom.paas
```
### 报错信息
```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'shiroFilterFactoryBean' defined in class path resource [com/freedom/paas/framework/authentication/config/ShiroConfig.class]: Failed to instantiate [org.apache.shiro.spring.web.ShiroFilterFactoryBean]: Factory method 'shiroFilterFactoryBean' threw exception with message: Failed to invoke the method queryList in the service com.freedom.paas.rpc.system.admin.service.AdminPermissionService. No provider available for the service com.freedom.paas.rpc.system.admin.service.AdminPermissionService from registry Directory(invokers: 1[127.0.0.1], validInvokers: 0[], invokersToReconnect: 1[127.0.0.1]) on the consumer 192.168.3.10 using the dubbo version 3.3.0-beta.1. Please check if the providers have been started and registered.
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:657) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:645) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1324) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1161) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:561) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:521) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:205) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.registerBeanPostProcessors(PostProcessorRegistrationDelegate.java:273) ~[spring-context-6.0.4.jar:6.0.4]
	at org.springframework.context.support.AbstractApplicationContext.registerBeanPostProcessors(AbstractApplicationContext.java:761) ~[spring-context-6.0.4.jar:6.0.4]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:568) ~[spring-context-6.0.4.jar:6.0.4]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.0.2.jar:3.0.2]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:730) ~[spring-boot-3.0.2.jar:3.0.2]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:432) ~[spring-boot-3.0.2.jar:3.0.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:308) ~[spring-boot-3.0.2.jar:3.0.2]
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:150) ~[spring-boot-3.0.2.jar:3.0.2]
	at com.freedom.paas.local.console.manager.NativeConsoleManagerMain.main(NativeConsoleManagerMain.java:19) ~[classes/:na]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.apache.shiro.spring.web.ShiroFilterFactoryBean]: Factory method 'shiroFilterFactoryBean' threw exception with message: Failed to invoke the method queryList in the service com.freedom.paas.rpc.system.admin.service.AdminPermissionService. No provider available for the service com.freedom.paas.rpc.system.admin.service.AdminPermissionService from registry Directory(invokers: 1[127.0.0.1], validInvokers: 0[], invokersToReconnect: 1[127.0.0.1]) on the consumer 192.168.3.10 using the dubbo version 3.3.0-beta.1. Please check if the providers have been started and registered.
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:171) ~[spring-beans-6.0.4.jar:6.0.4]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653) ~[spring-beans-6.0.4.jar:6.0.4]
	... 18 common frames omitted
Caused by: org.apache.dubbo.rpc.RpcException: Failed to invoke the method queryList in the service com.freedom.paas.rpc.system.admin.service.AdminPermissionService. No provider available for the service com.freedom.paas.rpc.system.admin.service.AdminPermissionService from registry Directory(invokers: 1[127.0.0.1], validInvokers: 0[], invokersToReconnect: 1[127.0.0.1]) on the consumer 192.168.3.10 using the dubbo version 3.3.0-beta.1. Please check if the providers have been started and registered.
	at org.apache.dubbo.rpc.cluster.support.AbstractClusterInvoker.checkInvokers(AbstractClusterInvoker.java:370) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.support.FailoverClusterInvoker.doInvoke(FailoverClusterInvoker.java:73) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.support.AbstractClusterInvoker.invoke(AbstractClusterInvoker.java:344) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.router.RouterSnapshotFilter.invoke(RouterSnapshotFilter.java:46) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.monitor.support.MonitorFilter.invoke(MonitorFilter.java:108) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.support.MetricsClusterFilter.invoke(MetricsClusterFilter.java:59) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.protocol.dubbo.filter.FutureFilter.invoke(FutureFilter.java:53) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.spring.security.filter.ContextHolderParametersSelectedTransferFilter.invoke(ContextHolderParametersSelectedTransferFilter.java:41) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.metrics.filter.MetricsFilter.invoke(MetricsFilter.java:76) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.support.MetricsConsumerFilter.invoke(MetricsConsumerFilter.java:37) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.support.ConsumerClassLoaderFilter.invoke(ConsumerClassLoaderFilter.java:40) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.tracing.filter.ObservationSenderFilter.invoke(ObservationSenderFilter.java:57) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.support.ConsumerContextFilter.invoke(ConsumerContextFilter.java:116) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CopyOfFilterChainNode.invoke(FilterChainBuilder.java:334) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.filter.FilterChainBuilder$CallbackRegistrationInvoker.invoke(FilterChainBuilder.java:196) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.support.wrapper.AbstractCluster$ClusterFilterInvoker.invoke(AbstractCluster.java:91) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.support.wrapper.MockClusterInvoker.invoke(MockClusterInvoker.java:103) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.cluster.support.wrapper.ScopeClusterInvoker.invoke(ScopeClusterInvoker.java:163) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.proxy.InvocationUtil.invoke(InvocationUtil.java:61) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at org.apache.dubbo.rpc.proxy.InvokerInvocationHandler.invoke(InvokerInvocationHandler.java:75) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at com.freedom.paas.rpc.system.admin.service.AdminPermissionServiceDubboProxy2.queryList(AdminPermissionServiceDubboProxy2.java) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.apache.dubbo.config.spring.util.LazyTargetInvocationHandler.invoke(LazyTargetInvocationHandler.java:58) ~[dubbo-3.3.0-beta.1.jar:3.3.0-beta.1]
	at com.freedom.paas.rpc.system.admin.service.AdminPermissionServiceDubboProxy2.queryList(AdminPermissionServiceDubboProxy2.java) ~[classes/:na]
	at com.freedom.paas.console.manager.system.admin.handler.SecurityManagerHandler.getAllPermission(SecurityManagerHandler.java:110) ~[classes/:na]
	at com.freedom.paas.framework.authentication.config.ShiroConfig.shiroFilterFactoryBean(ShiroConfig.java:88) ~[classes/:na]
	at com.freedom.paas.framework.authentication.config.ShiroConfig$$SpringCGLIB$$0.CGLIB$shiroFilterFactoryBean$3(<generated>) ~[classes/:na]
	at com.freedom.paas.framework.authentication.config.ShiroConfig$$SpringCGLIB$$2.invoke(<generated>) ~[classes/:na]
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258) ~[spring-core-6.0.4.jar:6.0.4]
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331) ~[spring-context-6.0.4.jar:6.0.4]
	at com.freedom.paas.framework.authentication.config.ShiroConfig$$SpringCGLIB$$0.shiroFilterFactoryBean(<generated>) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:139) ~[spring-beans-6.0.4.jar:6.0.4]
	... 19 common frames omitted

```

### PR DESCRIPTION
**Pull request by namelessssssssssss**
* fix  https://github.com/apache/dubbo/issues/13672

### CODE REVIEW

**Comment by namelessssssssssss** [comment]
@AlbumenJ PTAL

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ReferenceBean.java [hunk 1]
```diff
@@ -453,4 +460,17 @@ public Object getTarget() throws Exception {
...
+
+    private void tryExportLocalProvider() {
+        String providerBeanName = ServiceBeanNameBuilder.create(interfaceName, applicationContext.getEnvironment())
+                .group(getGroup())
+                .version(getVersion())
+                .build();
+        @SuppressWarnings("unchecked")
+        ServiceBean<T> serviceBean =
+                applicationContext.getBeansOfType(ServiceBean.class).get(providerBeanName);
+        if (serviceBean != null && !serviceBean.isExported() && SCOPE_LOCAL.equals(serviceBean.getScope())) {
```
Comment (namelessssssssssss) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ReferenceBean.java]: Should we add `SCOPE_LOCAL.equals(serviceBean.getScope())` ?

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ReferenceBean.java [hunk 2]
```diff
@@ -443,6 +445,11 @@ private Object getCallProxy() throws Exception {
             return referenceConfig.get();
         }
         synchronized (LockUtils.getSingletonMutex(applicationContext)) {
+            if (SCOPE_LOCAL.equals(referenceConfig.getScope())) {
+                // In some case, local(InJVM) providers are not exported when consumer calls
+                // like user call a local service when bean initializing
+                tryExportLocalProvider();
```
Comment (AlbumenJ) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ReferenceBean.java:451]: Export the service here is not a good idea. 
I think there might not exist an elegant resolution for such scenario.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Comment by ShenFeng312** [comment]
I think this issue doesn't need to be fixed. If this usage is really necessary, why not use  `@Service` instead?

**Pull request closed without merging**